src/Fork.php

```diff
@@ -80,26 +85,66 @@ protected function forkForTask(Task $task): Task
             ->setConnection($socketToChild);
     }
 
-    protected function waitFor(Task ...$runningTasks): array
+    protected function runTask(Task $task): Task
+    {
+        if ($this->toExecuteBeforeInParentTask) {
+            ($this->toExecuteBeforeInParentTask)();
+        }
+
+        return $this->forkForTask($task);
+    }
+
+    protected function finishTask(Task $task): mixed
+    {
+        $output = $task->output();
+
+        if ($this->toExecuteAfterInParentTask) {
+            ($this->toExecuteAfterInParentTask)($output);
+        }
+
+        return $output;
+    }
+
+    protected function waitFor(Task ...$queue): array
     {
         $output = [];
 
-        while (count($runningTasks)) {
-            foreach ($runningTasks as $key => $task) {
-                if ($task->isFinished()) {
-                    $taskOutput = $task->output();
+        $running = [];
+
+        $amountRunning = 0;
+
+        foreach ($queue as $i => $task) {
+            $running[] = $this->runTask($task);
+
+            unset($queue[$i]);
+
+            $amountRunning += 1;
+
+            if ($this->concurrent && $amountRunning >= $this->concurrent) {
```

Comment:
Let's extract this to a function with a good description, `maximumAmountOfTasksRunning` or better.